pkg/traces/traces.go

```diff
@@ -0,0 +1,115 @@
+	defaultSpanName           = "launcher/unknown"
+	defaultAttributeNamespace = "unknown"
+)
+
+// StartSpan returns a new context and span, including information about the calling function
+// as appropriate. Standardizes the tracer name. The caller is always responsible for
+// ending the span. `keyVals` should be a list of pairs, where the first in the pair is a
+// string representing the attribute key and the second in the pair is the attribute value.
+func StartSpan(ctx context.Context, keyVals ...interface{}) (context.Context, trace.Span) {
+	spanName := defaultSpanName
+
+	opts := make([]trace.SpanStartOption, 0)
+
+	// Extract information about the caller to set some standard attributes (code.filepath,
+	// code.lineno, code.function) and to set more specific span and attribute names.
+	// runtime.Caller(0) would return information about `StartSpan` -- calling
+	// runtime.Caller(1) will return information about the function calling `StartSpan`.
+	pc, callerFile, callerLine, ok := runtime.Caller(1)
+	if ok {
+		opts = append(opts, trace.WithAttributes(
+			semconv.CodeFilepath(callerFile),
+			semconv.CodeLineNumber(callerLine),
+		))
+
+		// Extract the calling function name and use it to set code.function and the span name.
+		if f := runtime.FuncForPC(pc); f != nil {
+			spanName = filepath.Base(f.Name())
+			opts = append(opts, trace.WithAttributes(semconv.CodeFunction(f.Name())))
+		}
+	}
+
+	opts = append(opts, trace.WithAttributes(buildAttributes(callerFile, keyVals...)...))
+
+	return otel.Tracer(instrumentationPkg).Start(ctx, spanName, opts...)
+}
+
+// SetError records the error on the span and sets the span's status to error.
+func SetError(span trace.Span, err error) {
+	span.RecordError(err)
+	span.SetStatus(codes.Error, err.Error())
+}
+
+// buildAttributes takes the given keyVals, expected to be pairs representing the key
+// and value of each attribute, and parses them appropriately, ensuring that the keys
+// have consistent and specific names. Pairs with invalid keys or values will be added
+// as string attributes.
+func buildAttributes(callerFile string, keyVals ...interface{}) []attribute.KeyValue {
+	callerDir := defaultAttributeNamespace
+	if callerFile != "" {
+		// This is the closest we get to grabbing the package name from the caller -- e.g. if
+		// the calling function is `/some/path/to/ee/localserver/request-controlservice.go`,
+		// this will extract `localserver`.
+		callerDir = filepath.Base(filepath.Dir(callerFile))
+	}
+
+	attrs := make([]attribute.KeyValue, 0)
+
+	for i := 0; i < len(keyVals); i += 2 {
+		// Keys must always be strings
+		if _, ok := keyVals[i].(string); !ok {
+			attrs = append(attrs, attribute.String(
+				fmt.Sprintf("bad key type %T: %v", keyVals[i], keyVals[i]),
+				fmt.Sprintf("%v", keyVals[i+1]),
+			))
+			continue
+		}
+
+		key := fmt.Sprintf("launcher.%s.%s", callerDir, keyVals[i])
+
+		// Create an attribute of the appropriate type
+		switch keyVals[i+1].(type) {
```

Comment:
```suggestion
		switch v := keyVals[i+1].(type) {
```

You can set `v` here, and not need all the additional typecasts below

Reply:
Oh yup that's much tidier, thank you! Updated